docs/dev/how-to/estimate-gas.md

```diff
@@ -0,0 +1,58 @@
+# Estimate gas
+
+:::tip Gas fees
+- Layer 2 gas fees depend on the current Ethereum L1 gas fees for publishing and verification.
+:::
+
+Gas estimation is usually left to the SDK to deal with. However, there are situations in which a developer may wish to implement gas estimation directly; such as for displaying the fee to the user for example.
+
+This document gives step-by-step instructions for implementing gas estimation for:
+
+- [L1 to L1](#l1-to-l1) transactions.
+- [L1 to L2](#l1-to-l2) transactions.
+- [L2 to L2](#l2-to-l2) transactions.
+
+## L1 to L1
+
+To estimate gas for L1 to L1 transactions, use the Ethereum `eth_estimateGas` method and send to an L1 node.
+
+:::info More info
+For more information and live testing, check out the [Ethereum JSON RPC docs](https://ethereum.github.io/execution-apis/api-documentation/).
+:::
+
+## L1 to L2
+
+To estimate gas for an L1 to L2 transaction, first gather the required values:
+
+1. Get the current L1 gas price. 
+    
+    For example, the Ethers.js library has a [`getGasPrice()`](https://docs.ethers.org/v5/api/providers/provider/#Provider-getGasPrice) method available on a `Provider` object. Your language of choice should have an equivalent.
+
+2. Call the [`zks_estimateGasL1ToL2`](../../api/api.md#zks_estimategasl1tol2) method, passing the transaction data. This returns the minimum amount of gas the transaction requires; commonly called **gas limit** or similar in our code and docs.
```

Comment:
That is not what I was told. Please advise @StanislavBreadless .. also, I'm going to leave in the commonly called bit because it was not clear that these variable names were referring to a gas estimation limit (minimum or maximum) ...

Reply:
I'll wait for Stas response and then commit the change ..

Reply:
> also, I'm going to leave in the commonly called bit because it was not clear that these variable names were referring to a gas estimation limit (minimum or maximum) ...

estimateGas should return the estimated amount gas to be consumed by the transaction. It should not be "gas limit" (whether min or max)

Reply:
The variable is nearly always called limit in the code, this is why I'm mentioning it, because it's darn confusing.

Reply:
i'm committing this change given the confusion

Reply:
> The variable is nearly always called (something containing) "limit" in our code, inc. libraries and examples, this is why I'm mentioning it, because it's darn confusing.

Just to be clear, there is a variable that is called gasLimit (and it's the max).
The process of assigning a value to it (how do we know what should be the max?) is to call the estimateGas function (note, a function, not variable). This gives us the amount of gas we expect the tx to consume (in EVM we cannot predict it 100%, hence only estimation). Once we have this estimation, we multiply it be some factor (1.1, 1.3, 2, whatever) just to make sure that we don't send a tx with too little of gas, which will result in tx failure. And this would be the gasLimit we assign to this tx.